src/wayback/_http.py

```diff
@@ -0,0 +1,282 @@
+            headers=raw.headers,
+            links=raw.links,
+            encoding=raw.encoding
+        )
+        self._read_lock = threading.RLock()
+        self._raw = raw
+
+    @property
+    def content(self) -> bytes:
+        with self._read_lock:
+            if self._content is None:
+                self._content = self._raw.content
+
+            return self._content
+
+    def sniff_encoding(self) -> None:
+        self.content
+        return self._raw.apparent_encoding
+
+    def close(self, cache: bool = True) -> None:
+        with self._read_lock:
+            # Reading bytes potentially involves decoding data from compressed
+            # gzip/brotli/etc. responses, so we need to handle those errors by
+            # continuing to just read the raw data off the socket instead.
+            #
+            # This fallback behavior is inspired by requests:
+            #   https://github.com/psf/requests/blob/eedd67462819f8dbf8c1c32e77f9070606605231/requests/sessions.py#L160-L163
+            # For urllib3, the appropriate errors to handle would be:
+            #   `(DecodeError, ProtocolError, RuntimeError)`
+            try:
+                if cache:
+                    try:
+                        self.content
+                    except (ChunkedEncodingError, ContentDecodingError, RuntimeError):
+                        self._raw.raw.read(decode_content=False)
+                else:
+                    self._raw.raw.read(decode_content=False)
+            finally:
+                self._raw.close()
+
+
+class WaybackHttpAdapter:
+    """
+    Handles making actual HTTP requests over the network. For now, this is an
+    implementation detail, but it may be a way for users to customize behavior
+    in the future.
+    """
+
+    def __init__(self) -> None:
+        self._session = requests.Session()
+
+    # FIXME: remove `allow_redirects`! Redirection needs to be handled by
+    # whatever does throttling, which is currently not here.
+    def request(
+        self,
+        method: str,
+        url: str,
+        *,
+        params: dict = None,
+        headers: dict = None,
+        allow_redirects: bool = True,
+        timeout: Union[int, Tuple[int, int]] = None
+    ) -> WaybackHttpResponse:
+        """
+        Send an HTTP request and return a :class:`WaybackHttpResponse` object
+        representing the response.
+
+        Parameters
+        ----------
+        method : str
+            Method to use for the request, e.g. ``'GET'``, ``'POST'``.
+        url : str
+            The URL to reqeust.
+        params : dict, optional
+            For POST/PUT requests, data to be encoded in the response body. For
+            other methods, this should be encoded as the querystring.
+        headers : dict, optional
+            The HTTP headers to send with the request.
+        allow_redirects : bool, default: True
+            Whether to follow redirects before returning a response.
```

Comment:
If we keep rate limiting out of the adapter layer, this argument should be removed, since redirect following has to be handled wherever rate limits are (or higher up).

Reply:
Rate limiting wound up back in the adapter layer, so this concern is no longer relevant.